File: test/human-python/demo-mcm-segment.jl
# demo_mcm_segment_minimal.jl
"""
Minimal MCM demo with segment-level temperature perturbation.

Instead of applying temperature uncertainty to the entire fiber,
we apply MCMadd(:T_K, ΔT) to a single bend segment via meta.
"""

using Bifrost
using Distributions
using MonteCarloMeasurements
using Printf

const N_SAMPLES = 100

# =============================
# Setup: fiber with multi-segment path
# =============================

xs = StepIndexCrossSection(
    SilicaGermaniaGlass(0.036),
    SilicaGermaniaGlass(0.0),
    8.2e-6,
    125.0e-6,
    manufacturer = "Corning",
    model_number = "SMF-like",
)

# Reference temperature: 24°C (fixed for the whole fiber)
T_ref_K = 297.15

# =============================
# MCM: Segment-level temperature perturbation
# =============================

# Temperature offset: ΔT ~ N(0, 5 K)  [±5°C variation around reference]
# This is passed to the bend via MCMadd(:T_K, ΔT_K)
ΔT_dist = Normal(0.0, 5.0)  # mean offset, std
ΔT_K_particles = Particles(N_SAMPLES, ΔT_dist)

println("Temperature perturbation ensemble:")
@printf "  Mean ΔT:  %.2f K\n" mean(ΔT_K_particles.particles)
@printf "  Std ΔT:   %.2f K\n" std(ΔT_K_particles.particles)
@printf "  Min ΔT:   %.2f K\n" minimum(ΔT_K_particles.particles)
@printf "  Max ΔT:   %.2f K\n" maximum(ΔT_K_particles.particles)

# Apply segment-level MCM perturbation via meta
# The bend segment carries MCMadd(:T_K, ΔT_K_particles) in its meta,
# which modify() will interpret as "add ΔT_K to the local temperature"
spec_with_meta = PathSpecBuilder()
straight!(spec_with_meta; length = 0.5, meta = [Nickname("lead-in")])
bend!(spec_with_meta; radius = 0.01, angle = π/2,
      meta = [Nickname("sensitive bend"), MCMadd(:T_K, ΔT_K_particles)])
straight!(spec_with_meta; length = 0.5, meta = [Nickname("lead-out")])

path_with_meta = build(spec_with_meta)
fiber_with_meta = Fiber(path_with_meta; cross_section = xs, T_ref_K = T_ref_K)

# =============================
# Modify: apply MCM perturbations to the path
# =============================

println("\nApplying MCM perturbations via modify()...")
modified_path = modify(fiber_with_meta)
fiber_modified = Fiber(modified_path; cross_section = xs, T_ref_K = T_ref_K)

# =============================
# Propagate
# =============================

println("Propagating with segment-level temperature variation...")
J_particles, stats = propagate_fiber(
    fiber_modified;
    λ_m = 1550e-9,
    rtol = 1e-9,
    verbose = false,
)

println("Done.\n")

# =============================
# Convert Jones matrix to rotation angle
# =============================

function jones_to_rotation_angle(J)
    tr_J = J[1,1] + J[2,2]
    cos_theta = real(tr_J) / 2
    cos_theta = clamp(cos_theta, -1.0, 1.0)
    return acos(cos_theta)
end

# =============================
# Extract per-sample angles
# =============================

println("Extracting rotation angles per sample...")
n_samples = length(real(J_particles[1, 1]).particles)
angles = Float64[]
T_eff = Float64[]  # Effective temperature for each sample

for k in 1:n_samples
    # Extract k-th sample
    J_k = ComplexF64[
        real(J_particles[1,1]).particles[k] + im*imag(J_particles[1,1]).particles[k]   real(J_particles[1,2]).particles[k] + im*imag(J_particles[1,2]).particles[k];
        real(J_particles[2,1]).particles[k] + im*imag(J_particles[2,1]).particles[k]   real(J_particles[2,2]).particles[k] + im*imag(J_particles[2,2]).particles[k]
    ]
    θ_k = jones_to_rotation_angle(J_k)
    push!(angles, θ_k)
    push!(T_eff, T_ref_K + ΔT_K_particles.particles[k])
end

# =============================
# Results
# =============================

@printf "\nResults (segment-level MCM perturbation):\n"
@printf "  Rotation angle (°):\n"
@printf "    Mean:  %.6f\n" mean(angles)*180/π
@printf "    Std:   %.6f\n" std(angles)*180/π
@printf "    Min:   %.6f\n" minimum(angles)*180/π
@printf "    Max:   %.6f\n" maximum(angles)*180/π

@printf "\nFirst 5 samples:\n"
for k in 1:min(5, n_samples)
    @printf "  Sample %d: T_eff=%.2f K (ΔT=%+.2f K), θ=%.6f°\n" k T_eff[k] ΔT_K_particles.particles[k] angles[k]*180/π
end

File: test/human-python/demo-mcm-segment.py
# demo_mcm_minimal.py
"""Minimal MCM demo: temperature-dependent Jones matrix rotation angle."""

import bifrost_py as bf
import numpy as np
from scipy.stats import norm

N_SAMPLES = 100


def jones_to_rotation_angle(J):
    """
    Extract rotation angle (rad) from a 2×2 Jones matrix.
    
    For a unitary matrix J = exp(i·θ·n̂·σ), the rotation angle θ
    satisfies: tr(J) = 2·cos(θ)
    
    Parameters
    ----------
    J : np.ndarray
        2×2 complex Jones matrix.
    
    Returns
    -------
    float
        Rotation angle in radians.
    """
    tr_J = np.trace(J)
    cos_theta = np.real(tr_J) / 2
    # Clamp to avoid numerical issues with arccos
    cos_theta = np.clip(cos_theta, -1.0, 1.0)
    return np.arccos(cos_theta)


# =============================
# Setup: fiber with uncertain temperature
# =============================

print("Setting up fiber...")

xs = bf.StepIndexCrossSection(
    bf.SilicaGermaniaGlass(0.036),
    bf.SilicaGermaniaGlass(0.0),
    8.2e-6,
    125.0e-6,
)


Delta_T_K = norm(loc=0.0, scale=5.0)

print("Creating Monte Carlo ensemble...")

# Temperature: 24°C ± 5°C (normal distribution, 50 samples)
T_ref_K = 297.15
Delta_T_K = norm(loc=0.0, scale=5.0)  # mean, std
Delta_T_K_particles = bf.mcm.StaticParticles(n=N_SAMPLES, distribution=Delta_T_K, seed=42)

print(f"  Temperature: {Delta_T_K_particles.mean:.2f} ± {Delta_T_K_particles.std:.2f} K")
print(f"  Samples: {Delta_T_K_particles.n}")

# Build a simple path: straight 1 meter
spec = bf.SubpathBuilder(); bf.start_b(spec)
bf.straight_b(spec, length = 0.5)
bf.bend_b(spec, radius = 0.01, angle = bf.pi / 2, meta=[bf.MCMadd('T_K', Delta_T_K_particles)])
bf.straight_b(spec, length = 0.5)
bf.seal_b(spec)
path = bf.build(spec)

# Reference temperature
T_ref_K = 297.15  # 24°C

# Fiber with uncertain temperature
fiber_mcm = bf.Fiber(path, cross_section=xs, T_ref_K=T_ref_K)
seg = spec.segments[1]  # Access the SECOND segment... Python indexing, not Julia
print("The bend's meta is ", seg.meta)

# =============================
# Propagate
# =============================

print("Propagating fiber with uncertain temperature...")
J_particles, stats = bf.propagate_fiber(
    fiber_mcm,
    lambda_m=1550e-9,
    verbose=False,
)

print(f"Done. J is a ParticlesMatrix with shape {J_particles.particles.shape}")
print(f"  (2, 2, {J_particles.n_samples}) samples\n")

# =============================
# Extract per-sample rotation angles
# =============================

print("Extracting rotation angles per sample...")

angles = []
for k in range(J_particles.n_samples):
    # Extract k-th sample: shape (2, 2)
    J_k = J_particles.particles[:, :, k]
    theta_k = jones_to_rotation_angle(J_k)
    angles.append(theta_k)

angles = np.array(angles)

# =============================
# Results
# =============================

print("\nResults:")
print(f"  Temperature range: {Delta_T_K_particles.particles.min():.2f}–{Delta_T_K_particles.particles.max():.2f} K")
print(f"  Rotation angle (°):")
print(f"    Mean:  {np.mean(angles*180/np.pi):.3f}")
print(f"    Std:   {np.std(angles*180/np.pi):.4f}")
print(f"    Min:   {np.min(angles*180/np.pi):.3f}")
print(f"    Max:   {np.max(angles*180/np.pi):.3f}")

print("\nFirst 5 samples:")
for k in range(min(5, J_particles.n_samples)):
    T_k = Delta_T_K_particles.particles[k]
    theta_k = angles[k]
    print(f"  Sample {k+1}: T={T_k-273.15:.2f}°C, θ={theta_k*180/np.pi:.3f}°")